### PR TITLE
V8 - Show that underlying item is used on the info tab

### LINKED ISSUE #8699: Show references of items being used in descendants on info tab
This is part of #8679

If a item it self does not have any references, we should check if one of it descendants has references. 
This is useful on for example folders in the media section. A editor can see which items in the folder are in use.

### PR DESCRIPTION
**Pull request by dawoe**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes #8699

### Description

With this PR you can now see if one of the underlying items (descendants) is being used in content. It doesn't show which content items yet, that is for a future PR.

To have good performance I needed to introduce a new repository and service method.

For testing the default starter kit can be used

### Steps to test for media

1. Create a folder structure several levels deep in the media section. 
2. Upload some photo's in a sub folder
3. Link one of the photo's from content
4. On the top level folder check the info tab. It should indicate that one of the underlying items is in use.

### Steps to test for content
1. Create a content structure several levels deep in the content section. 
2. Link content on the lowest level in another page.
3. On the top level content item of the newly created content structure check the info tab. It should indicate that one of the underlying items is in use.

### CODE REVIEW

**Comment by emmaburstow** [comment]
Nice work Dave, thanks for this. Will take a look and merge if all is well,

Em

**Comment by poornimanayar** [comment]
Hey @dawoe ,

Just so that I have got it right! I have attached gifs from my testing.

contentref
mediaref

**Comment by poornimanayar** [comment]
@Shazwazza Can you please cast an eye on this before it gets merged in? I have tested the work so far and it works well.

File: src/Umbraco.Web/Editors/TrackedReferencesController.cs [hunk 1]
```diff
@@ -57,5 +58,55 @@ public PagedResult<EntityBasic> GetPagedReferences(int id, string entityType, in
                 })
             };
         }
+
+        [HttpGet]
+        public bool HasReferencesInDescendants(int id, string entityType)
+        {
+            //this.Services.EntityService.GetChildren()
```
Comment (Shazwazza) [inline src/Umbraco.Web/Editors/TrackedReferencesController.cs:65]: do we need these comments?

File: src/Umbraco.Web/Editors/TrackedReferencesController.cs [hunk 2]
```diff
@@ -57,5 +58,55 @@ public PagedResult<EntityBasic> GetPagedReferences(int id, string entityType, in
...
+                var pageSize = 1000;
+                var currentPage = 0;
+
+                IEntitySlim[] entities;
+
+
+                do
+                {
+                    entities = this.Services.EntityService.GetPagedDescendants(id, currentObjectType, currentPage, pageSize, out _)
+                        .ToArray();
```
Comment (Shazwazza) [inline src/Umbraco.Web/Editors/TrackedReferencesController.cs:90]: ToList is faster than ToArray

File: src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html [hunk 3]
```diff
@@ -47,6 +47,18 @@ <h5 class="mt0" style="margin-bottom: 20px;">
...
     </div>
 
+    <div ng-if="vm.hasContentReferencesInDescendants">
+        <h5 class="mt0" style="margin-bottom: 20px;">
+            <localize key="references_labelUsedByDocuments">Used in Documents</localize>
+        </h5>
+
+        <umb-box>
+            <umb-box-content>
+                <localize key="references_labelUsedInContentDescendants">One of the underlying items is being used in content item</localize>
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html:57]: I'm not sure that 'underlying' is the correct term here, at least to me it seems odd and not sure we use that term for descendants anywhere else? Could it be "One or more of this item's descendants is being used in another content item" is that more specific to what this is trying to tell the user?

File: src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html [hunk 4]
```diff
@@ -83,6 +95,18 @@ <h5 class="mt0" style="margin-bottom: 20px;">
...
     </div>
 
+    <div ng-if="vm.hasMemberReferencesInDescendants">
+        <h5 class="mt0" style="margin-bottom: 20px;">
+            <localize key="references_labelUsedByMembers">Used in Members</localize>
+        </h5>
+
+        <umb-box>
+            <umb-box-content>
+                <localize key="references_labelUsedInMemberDescendants">One of the underlying items is being used in a member</localize>
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html:105]: as above

File: src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html [hunk 5]
```diff
@@ -119,4 +143,16 @@ <h5 class="mt0" style="margin-bottom: 20px;">
...
     </div>
 
+    <div ng-if="vm.hasMediaReferencesInDescendants">
+        <h5 class="mt0" style="margin-bottom: 20px;">
+            <localize key="references_labelUsedByMedia">Used in Media</localize>
+        </h5>
+
+        <umb-box>
+            <umb-box-content>
+                <localize key="references_labelUsedInMediaDescendants">One of the underlying items is being used in a media item</localize>
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html:153]: as above

File: src/Umbraco.Web.UI/Umbraco/config/lang/en.xml [hunk 6]
```diff
@@ -2292,6 +2292,9 @@ To manage your website, simply open the Umbraco back office and start adding con
     <key alias="labelUsedByDocuments">Used in Documents</key>
     <key alias="labelUsedByMembers">Used in Members</key>
     <key alias="labelUsedByMedia">Used in Media</key>
+      <key alias="labelUsedInMediaDescendants">One of the underlying items is being used in a media item</key>
+      <key alias="labelUsedInContentDescendants">One of the underlying items is being used in a content item</key>
+      <key alias="labelUsedInMemberDescendants">One of the underlying items is being used in a member</key>
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI/Umbraco/config/lang/en.xml:2297]: as above

File: src/Umbraco.Web.UI/Umbraco/config/lang/en_us.xml [hunk 7]
```diff
@@ -2314,6 +2314,9 @@ To manage your website, simply open the Umbraco back office and start adding con
     <key alias="labelUsedByDocuments">Used in Documents</key>
     <key alias="labelUsedByMembers">Used in Members</key>
     <key alias="labelUsedByMedia">Used in Media</key>
+      <key alias="labelUsedInMediaDescendants">One of the underlying items is being used in a media item</key>
+      <key alias="labelUsedInContentDescendants">One of the underlying items is being used in a content item</key>
+      <key alias="labelUsedInMemberDescendants">One of the underlying items is being used in a member</key>
```
Comment (Shazwazza) [inline src/Umbraco.Web.UI/Umbraco/config/lang/en_us.xml:2319]: as above

File: src/Umbraco.Core/Services/Implement/RelationService.cs [hunk 8]
```diff
@@ -294,6 +294,17 @@ public IEnumerable<IUmbracoEntity> GetPagedParentEntitiesByChildId(int id, long
             }
         }
 
+        public IEnumerable<IUmbracoEntity> GetPagedParentEntitiesByChildIds(int[] ids, long pageIndex, int pageSize, out long totalChildren,
```
Comment (Shazwazza) [inline src/Umbraco.Core/Services/Implement/RelationService.cs:297]: Can we have a unit test for this one? Should be pretty straight forward to use the other RelationServiceTests methods as a starting point.

**Review by Shazwazza — CHANGES_REQUESTED** [review]
Have left some comments but otherwise looks good. (PS I'm on holidays for 3 weeks so if everyone is happy with the changes that get made feel free to accept)

File: src/Umbraco.Web/Editors/TrackedReferencesController.cs (hunk 1 above)
Comment (dawoe) [reply src/Umbraco.Web/Editors/TrackedReferencesController.cs]: Removed them

File: src/Umbraco.Web/Editors/TrackedReferencesController.cs (hunk 2 above)
Comment (dawoe) [reply src/Umbraco.Web/Editors/TrackedReferencesController.cs]: Fixed as well

File: src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html (hunk 3 above)
Comment (dawoe) [reply src/Umbraco.Web.UI.Client/src/views/components/references/umb-tracked-references.html]: Update the translations

File: src/Umbraco.Core/Services/Implement/RelationService.cs (hunk 8 above)
Comment (dawoe) [reply src/Umbraco.Core/Services/Implement/RelationService.cs:297]: I added a unit test for this one. I noticed a lot more methods on this Service are untested. But that is maybe something for another PR

**Review by poornimanayar — APPROVED** [review]

**Comment by poornimanayar** [comment]
Thanks for the PR @dawoe This has now been merged

**Pull request merged**